atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosTimestampBoundStore.java

```diff
@@ -0,0 +1,214 @@
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.concurrent.ExecutorService;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
+import org.immutables.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Ordering;
+import com.palantir.atlasdb.encoding.PtBytes;
+import com.palantir.common.concurrent.PTExecutors;
+import com.palantir.common.remoting.ServiceNotAvailableException;
+import com.palantir.paxos.PaxosAcceptor;
+import com.palantir.paxos.PaxosLearner;
+import com.palantir.paxos.PaxosProposer;
+import com.palantir.paxos.PaxosQuorumChecker;
+import com.palantir.paxos.PaxosResponse;
+import com.palantir.paxos.PaxosRoundFailureException;
+import com.palantir.paxos.PaxosValue;
+import com.palantir.timestamp.MultipleRunningTimestampServiceError;
+import com.palantir.timestamp.TimestampBoundStore;
+
+public class PaxosTimestampBoundStore implements TimestampBoundStore {
+    private static final Logger log = LoggerFactory.getLogger(PaxosTimestampBoundStore.class);
+
+    private final PaxosProposer proposer;
+    private final PaxosLearner knowledge;
+
+    private final List<PaxosAcceptor> acceptors;
+    private final List<PaxosLearner> learners;
+    private final long randomWaitBeforeProposalMs;
+
+    @GuardedBy("this")
+    private SequenceAndBound agreedState;
+
+    private final ExecutorService executor = PTExecutors.newCachedThreadPool(PTExecutors.newNamedThreadFactory(true));
+
+    public PaxosTimestampBoundStore(PaxosProposer proposer,
+            PaxosLearner knowledge,
+            List<PaxosAcceptor> acceptors,
+            List<PaxosLearner> learners,
+            long randomWaitBeforeProposalMs) {
+        this.proposer = proposer;
+        this.knowledge = knowledge;
+        this.acceptors = acceptors;
+        this.learners = learners;
+        this.randomWaitBeforeProposalMs = randomWaitBeforeProposalMs;
+    }
+
+    @Override
+    public synchronized long getUpperLimit() {
+        List<PaxosLong> responses = PaxosQuorumChecker.collectQuorumResponses(
+                ImmutableList.copyOf(acceptors),
+                a -> ImmutablePaxosLong.of(a.getLatestSequencePreparedOrAccepted()),
+                proposer.getQuorumSize(),
+                executor,
+                PaxosQuorumChecker.DEFAULT_REMOTE_REQUESTS_TIMEOUT_IN_SECONDS,
+                true);
+        if (!PaxosQuorumChecker.hasQuorum(responses, proposer.getQuorumSize())) {
+            throw new ServiceNotAvailableException("could not get a quorum");
+        }
+        PaxosLong max = Ordering.natural().onResultOf(PaxosLong::getValue).max(responses);
+        agreedState = getAgreedState(max.getValue());
+        return agreedState.getBound();
+    }
+
+    @VisibleForTesting
```

Comment:
This class is pretty long, and there is clearly testable behaviour here. I lack syntax highlighting here, but perhaps we can pull out a class (or classes) with public (maybe static?) methods that we can test.